File: src/backend/api/services/quality_service.py
"""
Service layer for data quality operations in the self-healing data pipeline.
Provides functionality for managing quality rules, executing validations,
calculating quality scores, and handling quality issues. Acts as an
intermediary between API controllers and the underlying quality framework.
"""

import typing
from typing import List, Dict, Optional, Any, Tuple
import datetime
import uuid
import pandas  # version 2.0.x

from src.backend.constants import (  # src/backend/constants.py
    ValidationStatus,
    QualityDimension
)
from src.backend.config import get_config  # src/backend/config.py
from src.backend.utils.logging.logger import get_logger  # src/backend/utils/logging/logger.py
from src.backend.db.repositories.quality_repository import QualityRepository  # src/backend/db/repositories/quality_repository.py
from src.backend.quality.engines.validation_engine import ValidationEngine, ValidationResult, ValidationSummary  # src/backend/quality/engines/validation_engine.py
from src.backend.quality.engines.quality_scorer import QualityScorer, ScoringModel  # src/backend/quality/engines/quality_scorer.py
from src.backend.utils.storage.bigquery_client import BigQueryClient  # src/backend/utils/storage/bigquery_client.py
from src.backend.api.models.data_models import QualityRule, QualityValidation  # src/backend/api/models/data_models.py
from src.backend.api.models.error_models import ResourceNotFoundError, ValidationError  # src/backend/api/models/error_models.py
from src.backend.self_healing.ai.issue_classifier import classify_quality_issue  # src/backend/self_healing/ai/issue_classifier.py

# Initialize logger
logger = get_logger(__name__)

# Define global constants
DEFAULT_QUALITY_THRESHOLD = 0.9
DEFAULT_PAGE_SIZE = 20


def get_quality_rules(
    page: int,
    page_size: int,
    target_dataset: Optional[str] = None,
    target_table: Optional[str] = None,
    rule_type: Optional[str] = None,
    is_active: Optional[bool] = None
) -> Tuple[List[QualityRule], int]:
    """Retrieves a paginated list of quality rules with optional filtering"""
    logger.info(f"Request for quality rules with filters: page={page}, page_size={page_size}, "
                f"target_dataset={target_dataset}, target_table={target_table}, rule_type={rule_type}, is_active={is_active}")

    bq_client = BigQueryClient()
    repository = QualityRepository(bq_client=bq_client)

    page = page if page is not None else 1
    page_size = page_size if page_size is not None else DEFAULT_PAGE_SIZE

    offset = (page - 1) * page_size

    if target_dataset or target_table:
        rules = repository.get_rules_by_target(target_dataset, target_table, offset=offset, limit=page_size, is_active=is_active)
    elif rule_type:
        rules = repository.get_rules_by_type(rule_type, offset=offset, limit=page_size, is_active=is_active)
    else:
        rules = repository.get_all_rules(offset=offset, limit=page_size, is_active=is_active)

    total_count = repository.get_total_rules_count(target_dataset=target_dataset, target_table=target_table,
                                                    rule_type=rule_type, is_active=is_active)

    return rules, total_count


def get_quality_rule_by_id(rule_id: str) -> Optional[QualityRule]:
    """Retrieves a specific quality rule by ID"""
    logger.info(f"Request for quality rule by ID: {rule_id}")

    bq_client = BigQueryClient()
    repository = QualityRepository(bq_client=bq_client)

    rule = repository.get_rule_by_id(rule_id)

    return rule


def create_quality_rule(rule_data: QualityRule) -> QualityRule:
    """Creates a new quality rule"""
    logger.info(f"Request to create quality rule: {rule_data.rule_name}")

    # Validate required fields in rule_data
    if not rule_data.rule_name or not rule_data.target_dataset or not rule_data.target_table or not rule_data.rule_type:
        raise ValidationError("Missing required fields in rule data")

    bq_client = BigQueryClient()
    repository = QualityRepository(bq_client=bq_client)

    # Generate rule_id if not provided
    if not rule_data.rule_id:
        rule_data.rule_id = str(uuid.uuid4())

    # Set created_at and updated_at timestamps
    rule_data.created_at = datetime.datetime.now()
    rule_data.updated_at = datetime.datetime.now()

    # Call create_rule on repository
    created_rule = repository.create_rule(rule_data)

    # Retrieve the created rule to return complete object
    retrieved_rule = repository.get_rule_by_id(created_rule.rule_id)

    return retrieved_rule


def update_quality_rule(rule_id: str, rule_update: QualityRule) -> QualityRule:
    """Updates an existing quality rule"""
    logger.info(f"Request to update quality rule: {rule_id}")

    bq_client = BigQueryClient()
    repository = QualityRepository(bq_client=bq_client)

    # Get existing rule by ID
    existing_rule = repository.get_rule_by_id(rule_id)
    if not existing_rule:
        raise ResourceNotFoundError(resource_type="QualityRule", resource_id=rule_id)

    # Update rule fields with values from rule_update
    existing_rule.rule_name = rule_update.rule_name
    existing_rule.target_dataset = rule_update.target_dataset
    existing_rule.target_table = rule_update.target_table
    existing_rule.rule_type = rule_update.rule_type
    existing_rule.expectation_type = rule_update.expectation_type
    existing_rule.rule_definition = rule_update.rule_definition
    existing_rule.severity = rule_update.severity
    existing_rule.is_active = rule_update.is_active
    existing_rule.description = rule_update.description
    existing_rule.updated_at = datetime.datetime.now()
    existing_rule.metadata = rule_update.metadata

    # Call update_rule on repository
    updated_rule = repository.update_rule(existing_rule)

    # Retrieve the updated rule to return complete object
    retrieved_rule = repository.get_rule_by_id(updated_rule.rule_id)

    return retrieved_rule


def delete_quality_rule(rule_id: str) -> bool:
    """Deletes a quality rule"""
    logger.info(f"Request to delete quality rule: {rule_id}")

    bq_client = BigQueryClient()
    repository = QualityRepository(bq_client=bq_client)

    # Get existing rule by ID
    existing_rule = repository.get_rule_by_id(rule_id)
    if not existing_rule:
        raise ResourceNotFoundError(resource_type="QualityRule", resource_id=rule_id)

    # Call delete_rule on repository
    success = repository.delete_rule(existing_rule.rule_id)

    return success


def get_quality_validations(
    page: int,
    page_size: int,
    start_date: Optional[datetime.datetime] = None,
    end_date: Optional[datetime.datetime] = None,
    execution_id: Optional[str] = None,
    rule_id: Optional[str] = None,
    status: Optional[ValidationStatus] = None
) -> Tuple[List[QualityValidation], int]:
    """Retrieves a paginated list of quality validations with optional filtering"""
    logger.info(f"Request for quality validations with filters: page={page}, page_size={page_size}, "
                f"start_date={start_date}, end_date={end_date}, execution_id={execution_id}, rule_id={rule_id}, status={status}")

    bq_client = BigQueryClient()
    repository = QualityRepository(bq_client=bq_client)

    page = page if page is not None else 1
    page_size = page_size if page_size is not None else DEFAULT_PAGE_SIZE

    offset = (page - 1) * page_size

    search_criteria = {}
    if start_date:
        search_criteria['start_date'] = start_date
    if end_date:
        search_criteria['end_date'] = end_date
    if execution_id:
        search_criteria['execution_id'] = execution_id
    if rule_id:
        search_criteria['rule_id'] = rule_id
    if status:
        search_criteria['status'] = status.value

    validations, total_count = repository.search_validations(search_criteria, offset=offset, limit=page_size)

    return validations, total_count


def get_validation_by_id(validation_id: str) -> Optional[QualityValidation]:
    """Retrieves a specific validation result by ID"""
    logger.info(f"Request for validation by ID: {validation_id}")

    bq_client = BigQueryClient()
    repository = QualityRepository(bq_client=bq_client)

    validation = repository.get_validation_by_id(validation_id)

    return validation


def execute_validation(
    dataset: str,
    table: Optional[str] = None,
    execution_id: Optional[str] = None,
    rule_ids: Optional[List[str]] = None
) -> QualityValidation:
    """Executes quality validation for a dataset"""
    logger.info(f"Request to execute validation for dataset: {dataset}, table: {table}, "
                f"execution_id={execution_id}, rule_ids={rule_ids}")

    bq_client = BigQueryClient()
    repository = QualityRepository(bq_client=bq_client)
    validation_engine = ValidationEngine(config=get_config())

    # Get rules to validate (all applicable or specified by rule_ids)
    if rule_ids:
        rules = [repository.get_rule_by_id(rule_id) for rule_id in rule_ids]
    else:
        rules = repository.get_all_rules(target_dataset=dataset, target_table=table)

    # Prepare dataset for validation (load from BigQuery)
    # (Implementation depends on dataset type)
    # For now, assume dataset is a BigQuery table
    dataset_for_validation = bq_client.get_table(f"{dataset}.{table}")

    # Execute validation using ValidationEngine
    validation_summary = validation_engine.validate(dataset_for_validation, rules)

    # Process validation results
    # (Implementation depends on validation framework)
    # For now, just log the results
    logger.info(f"Validation results: {validation_summary}")

    # Store validation results in repository
    # (Implementation depends on repository)
    # For now, just log the results
    logger.info("Storing validation results in repository")

    # Trigger self-healing for failed validations if enabled
    # (Implementation depends on self-healing engine)
    # For now, just log the action
    logger.info("Triggering self-healing for failed validations if enabled")

    # Return validation summary
    return validation_summary


def get_quality_score(
    dataset: str,
    table: Optional[str] = None,
    as_of_date: Optional[datetime.datetime] = None
) -> Dict[str, Any]:
    """Retrieves the quality score for a dataset"""
    logger.info(f"Request for quality score for dataset: {dataset}, table: {table}, as_of_date={as_of_date}")

    bq_client = BigQueryClient()
    repository = QualityRepository(bq_client=bq_client)

    # Prepare filters for dataset, table, and date
    filters = {
        "dataset": dataset,
        "table": table,
        "as_of_date": as_of_date
    }

    # Call get_quality_metrics on repository
    quality_metrics = repository.get_quality_metrics(filters)

    # Format quality score response with overall score and dimension scores
    # (Implementation depends on scoring model)
    # For now, just return the raw metrics
    logger.info(f"Quality score: {quality_metrics}")
    return quality_metrics


def get_quality_issues(
    page: int,
    page_size: int,
    dataset: str,
    table: Optional[str] = None,
    severity: Optional[str] = None,
    is_resolved: Optional[bool] = None,
    start_date: Optional[datetime.datetime] = None,
    end_date: Optional[datetime.datetime] = None
) -> Tuple[List[Dict[str, Any]], int]:
    """Retrieves quality issues for a dataset"""
    logger.info(f"Request for quality issues: dataset={dataset}, table={table}, severity={severity}, "
                f"is_resolved={is_resolved}, start_date={start_date}, end_date={end_date}")

    bq_client = BigQueryClient()
    repository = QualityRepository(bq_client=bq_client)

    page = page if page is not None else 1
    page_size = page_size if page_size is not None else DEFAULT_PAGE_SIZE

    offset = (page - 1) * page_size

    # Prepare search criteria for failed validations
    search_criteria = {
        "dataset": dataset,
        "table": table,
        "severity": severity,
        "is_resolved": is_resolved,
        "start_date": start_date,
        "end_date": end_date
    }

    # Get failed validations from repository
    failed_validations = repository.get_failed_validations(search_criteria, offset=offset, limit=page_size)

    # Transform validation failures into quality issues with additional context
    # (Implementation depends on data model and context enrichment)
    # For now, just return the raw validation failures
    logger.info(f"Found {len(failed_validations)} quality issues")

    # Classify issues using self_healing.ai.issue_classifier if available
    # (Implementation depends on AI model integration)
    # For now, just log the action
    logger.info("Classifying issues using self_healing.ai.issue_classifier if available")

    # Return tuple of (issues list, total count)
    return failed_validations, len(failed_validations)


def get_validation_trend(
    dataset: str,
    table: Optional[str] = None,
    interval: str = "day",
    num_intervals: int = 30
) -> Dict[str, Any]:
    """Retrieves validation pass/fail trend over time"""
    logger.info(f"Request for validation trend: dataset={dataset}, table={table}, interval={interval}, num_intervals={num_intervals}")

    bq_client = BigQueryClient()
    repository = QualityRepository(bq_client=bq_client)

    # Validate interval parameter (hourly, daily, weekly)
    if interval not in ["hourly", "daily", "weekly"]:
        raise ValueError("Invalid interval: must be hourly, daily, or weekly")

    # Prepare filters for dataset and table
    filters = {
        "dataset": dataset,
        "table": table
    }

    # Call get_validation_trend on repository
    trend_data = repository.get_validation_trend(filters, interval, num_intervals)

    # Format trend data for response
    # (Implementation depends on data model and visualization requirements)
    # For now, just return the raw trend data
    logger.info(f"Validation trend data: {trend_data}")
    return trend_data


def trigger_self_healing(failed_validations: List[ValidationResult], execution_id: str) -> Dict[str, Any]:
    """Triggers self-healing for failed validations"""
    logger.info(f"Request to trigger self-healing for {len(failed_validations)} validations, execution_id={execution_id}")

    # Check if self-healing is enabled in configuration
    config = get_config()
    if not config.get("self_healing.enabled", False):
        logger.info("Self-healing is disabled in configuration")
        return {"status": "disabled"}

    # For each failed validation, classify issue type
    # (Implementation depends on AI model integration)
    # For now, just log the action
    logger.info("Classifying issue types for failed validations")

    # Determine appropriate healing action based on issue type
    # (Implementation depends on AI model integration and action mapping)
    # For now, just log the action
    logger.info("Determining appropriate healing actions")

    # Apply healing action if confidence above threshold
    # (Implementation depends on self-healing engine and action execution)
    # For now, just log the action
    logger.info("Applying healing actions if confidence above threshold")

    # Record healing attempt in validation record
    # (Implementation depends on repository)
    # For now, just log the action
    logger.info("Recording healing attempts in validation records")

    # Collect healing results for reporting
    # (Implementation depends on self-healing engine and reporting)
    # For now, just log the action
    logger.info("Collecting healing results for reporting")

    # Return healing results dictionary
    return {"status": "success"}


def convert_validation_to_api_model(validation_result: ValidationResult, validation_summary: ValidationSummary) -> QualityValidation:
    """Converts internal validation result to API model"""
    # Create new QualityValidation instance
    quality_validation = QualityValidation(
        validation_id=validation_result.validation_id,
        rule_id=validation_result.rule_id,
        execution_id=validation_result.execution_id,
        status=validation_result.status
    )

    # Map ValidationResult fields to QualityValidation fields
    quality_validation.success_percent = validation_summary.success_rate
    quality_validation.records_validated = validation_summary.total_rules
    quality_validation.records_failed = validation_summary.failed_rules
    quality_validation.validation_results = validation_result.details

    # Add summary information from ValidationSummary
    quality_validation.validation_time = datetime.datetime.now()  # Replace with actual validation time
    quality_validation.metadata = {"quality_score": validation_summary.quality_score}

    # Format validation details for API consumption
    # (Implementation depends on API requirements)
    # For now, just return the QualityValidation object
    return quality_validation


def convert_rule_to_api_model(rule_data: Dict[str, Any]) -> QualityRule:
    """Converts internal rule model to API model"""
    # Create new QualityRule instance
    quality_rule = QualityRule(
        rule_id=rule_data["rule_id"],
        rule_name=rule_data["rule_name"],
        target_dataset=rule_data["target_dataset"],
        target_table=rule_data["target_table"],
        rule_type=rule_data["rule_type"],
        expectation_type=rule_data["expectation_type"],
        rule_definition=rule_data["rule_definition"],
        severity=rule_data["severity"]
    )

    # Format rule definition for API consumption
    # (Implementation depends on API requirements)
    # For now, just return the QualityRule object
    return quality_rule


class QualityService:
    """Service class for data quality operations"""

    def __init__(self, bq_client: Optional[BigQueryClient] = None, config: Optional[Dict[str, Any]] = None):
        """Initializes the QualityService with necessary components"""
        # Load configuration from config parameter or get_config()
        self._config = config if config is not None else get_config()

        # Initialize BigQueryClient if not provided
        self._bq_client = bq_client if bq_client is not None else BigQueryClient()

        # Initialize QualityRepository with BigQueryClient
        self._repository = QualityRepository(bq_client=self._bq_client)

        # Initialize ValidationEngine with configuration
        self._validation_engine = ValidationEngine(config=self._config)

        logger.info("QualityService initialized")

    def get_rules(self, page: int, page_size: int, filters: Optional[Dict[str, Any]] = None) -> Tuple[List[QualityRule], int]:
        """Retrieves quality rules with pagination and filtering"""
        # Set default page and page_size if not provided
        page = page if page is not None else 1
        page_size = page_size if page_size is not None else DEFAULT_PAGE_SIZE

        # Extract filter parameters from filters dictionary
        target_dataset = filters.get("target_dataset") if filters else None
        target_table = filters.get("target_table") if filters else None
        rule_type = filters.get("rule_type") if filters else None
        is_active = filters.get("is_active") if filters else None

        # Call appropriate repository method based on filters
        rules, total_count = get_quality_rules(page, page_size, target_dataset, target_table, rule_type, is_active)

        # Convert repository models to API models
        api_rules = [convert_rule_to_api_model(rule) for rule in rules]

        # Return tuple of (rules list, total count)
        return api_rules, total_count

    def get_rule_by_id(self, rule_id: str) -> Optional[QualityRule]:
        """Retrieves a specific quality rule by ID"""
        # Call get_rule_by_id on repository
        rule = get_quality_rule_by_id(rule_id)

        # If rule found, convert to API model
        if rule:
            api_rule = convert_rule_to_api_model(rule)
            return api_rule

        # Return the rule or None if not found
        return None

    def create_rule(self, rule_data: QualityRule) -> QualityRule:
        """Creates a new quality rule"""
        # Validate required fields in rule_data
        # Convert API model to repository model
        # Call create_rule on repository
        created_rule = create_quality_rule(rule_data)

        # Retrieve the created rule
        # Convert back to API model and return
        return created_rule

    def update_rule(self, rule_id: str, rule_update: QualityRule) -> QualityRule:
        """Updates an existing quality rule"""
        # Get existing rule by ID
        # If rule not found, raise ResourceNotFoundError
        # Update rule fields with values from rule_update
        # Call update_rule on repository
        updated_rule = update_quality_rule(rule_id, rule_update)

        # Retrieve the updated rule
        # Convert to API model and return
        return updated_rule

    def delete_rule(self, rule_id: str) -> bool:
        """Deletes a quality rule"""
        # Get existing rule by ID
        # If rule not found, raise ResourceNotFoundError
        # Call delete_rule on repository
        success = delete_quality_rule(rule_id)

        # Return success status
        return success

    def get_validations(self, page: int, page_size: int, filters: Optional[Dict[str, Any]] = None) -> Tuple[List[QualityValidation], int]:
        """Retrieves quality validations with pagination and filtering"""
        # Set default page and page_size if not provided
        # Extract filter parameters from filters dictionary
        start_date = filters.get("start_date") if filters else None
        end_date = filters.get("end_date") if filters else None
        execution_id = filters.get("execution_id") if filters else None
        rule_id = filters.get("rule_id") if filters else None
        status = filters.get("status") if filters else None

        # Prepare search criteria with filters
        # Call search_validations on repository
        validations, total_count = get_quality_validations(page, page_size, start_date, end_date, execution_id, rule_id, status)

        # Convert repository models to API models
        # Return tuple of (validations list, total count)
        return validations, total_count

    def get_validation_by_id(self, validation_id: str) -> Optional[QualityValidation]:
        """Retrieves a specific validation by ID"""
        # Call get_validation_by_id on repository
        validation = get_validation_by_id(validation_id)

        # If validation found, convert to API model
        # Return the validation or None if not found
        return validation

    def execute_validation(
        self,
        dataset: str,
        table: Optional[str] = None,
        execution_id: Optional[str] = None,
        rule_ids: Optional[List[str]] = None
    ) -> QualityValidation:
        """Executes quality validation for a dataset"""
        # Prepare dataset for validation (load from BigQuery)
        # Get rules to validate (all applicable or specified by rule_ids)
        # Execute validation using ValidationEngine
        validation_summary = execute_validation(dataset, table, execution_id, rule_ids)

        # Process validation results
        # Trigger self-healing for failed validations if enabled
        # Convert results to API model and return
        return validation_summary

    def get_quality_score(
        self,
        dataset: str,
        table: Optional[str] = None,
        as_of_date: Optional[datetime.datetime] = None
    ) -> Dict[str, Any]:
        """Calculates quality score for a dataset"""
        # Prepare filters for dataset, table, and date
        # Call get_quality_metrics on repository
        quality_score = get_quality_score(dataset, table, as_of_date)

        # Format quality score response
        # Return quality score details dictionary
        return quality_score

    def get_quality_issues(
        self,
        page: int,
        page_size: int,
        dataset: str,
        table: Optional[str] = None,
        filters: Optional[Dict[str, Any]] = None
    ) -> Tuple[List[Dict[str, Any]], int]:
        """Retrieves quality issues for a dataset"""
        # Set default page and page_size if not provided
        # Extract filter parameters from filters dictionary
        severity = filters.get("severity") if filters else None
        is_resolved = filters.get("is_resolved") if filters else None
        start_date = filters.get("start_date") if filters else None
        end_date = filters.get("end_date") if filters else None

        # Prepare search criteria for failed validations
        # Get failed validations from repository
        issues, total_count = get_quality_issues(page, page_size, dataset, table, severity, is_resolved, start_date, end_date)

        # Transform validation failures into quality issues
        # Classify issues using AI if available
        # Apply additional filters (severity, resolution status)
        # Return tuple of (issues list, total count)
        return issues, total_count

    def get_validation_trend(
        self,
        dataset: str,
        table: Optional[str] = None,
        interval: str = "day",
        num_intervals: int = 30
    ) -> Dict[str, Any]:
        """Retrieves validation trend over time"""
        # Validate interval parameter
        # Prepare filters for dataset and table
        # Call get_validation_trend on repository
        trend_data = get_validation_trend(dataset, table, interval, num_intervals)

        # Format trend data for response
        # Return trend data dictionary
        return trend_data

    def trigger_self_healing(self, failed_validations: List[ValidationResult], execution_id: str) -> Dict[str, Any]:
        """Triggers self-healing for failed validations"""
        # Check if self-healing is enabled
        # For each failed validation, classify issue
        # Determine and apply healing action
        # Record healing attempt in validation
        # Return healing results
        return trigger_self_healing(failed_validations, execution_id)

    def close(self) -> None:
        """Closes the service and releases resources"""
        # Close validation engine if initialized
        if self._validation_engine:
            self._validation_engine.close()

        # Close BigQuery client if owned by this service
        if self._bq_client:
            self._bq_client.close()

        logger.info("QualityService closed")

File: src/web/src/services/api/qualityService.ts
/**
 * Quality Service
 * 
 * Service module for interacting with data quality-related API endpoints in the self-healing 
 * data pipeline application. Provides functions for retrieving and managing quality rules,
 * validation results, quality scores, and quality issues.
 */

import apiClient from './apiClient';
import { endpoints } from '../../config/apiConfig';
import {
  QualityRule,
  QualityValidation,
  QualityScore,
  ValidationStatus,
  PaginationParams,
  DateRangeParams,
  DataResponse,
  ListResponse
} from '../../types/api';
import {
  DatasetQualitySummary,
  QualityIssue,
  QualityRuleDetails,
  QualityValidationResult,
  QualityTimeSeries,
  ColumnQualityDetails,
  QualityStatistics,
  QualityDashboardFilters,
  QualityRuleTemplate,
  QualityDimension,
  QualityRuleType,
  QualityIssueStatus
} from '../../types/quality';
import { logError } from '../../utils/errorHandling';

/**
 * Retrieves a paginated list of quality rules with optional filtering
 * @param params - Pagination and filter parameters
 * @returns Promise resolving to a paginated list of quality rules
 */
const getQualityRules = async (
  params: PaginationParams & { 
    dataset?: string, 
    table?: string, 
    ruleType?: QualityRuleType, 
    isActive?: boolean 
  }
): Promise<ListResponse<QualityRule>> => {
  try {
    return await apiClient.get<ListResponse<QualityRule>>(
      endpoints.quality.rules,
      { params }
    );
  } catch (error) {
    logError(error, 'getQualityRules');
    throw error;
  }
};

/**
 * Retrieves a single quality rule by ID
 * @param ruleId - ID of the quality rule to retrieve
 * @returns Promise resolving to the detailed quality rule data
 */
const getQualityRule = async (
  ruleId: string
): Promise<DataResponse<QualityRuleDetails>> => {
  try {
    return await apiClient.get<DataResponse<QualityRuleDetails>>(
      endpoints.quality.ruleDetails(ruleId)
    );
  } catch (error) {
    logError(error, 'getQualityRule');
    throw error;
  }
};

/**
 * Creates a new quality rule
 * @param ruleData - Data for the new quality rule
 * @returns Promise resolving to the created quality rule
 */
const createQualityRule = async (
  ruleData: Partial<QualityRule>
): Promise<DataResponse<QualityRule>> => {
  try {
    return await apiClient.post<DataResponse<QualityRule>>(
      endpoints.quality.rules,
      ruleData
    );
  } catch (error) {
    logError(error, 'createQualityRule');
    throw error;
  }
};

/**
 * Updates an existing quality rule
 * @param ruleId - ID of the quality rule to update
 * @param ruleData - Updated rule data
 * @returns Promise resolving to the updated quality rule
 */
const updateQualityRule = async (
  ruleId: string,
  ruleData: Partial<QualityRule>
): Promise<DataResponse<QualityRule>> => {
  try {
    return await apiClient.put<DataResponse<QualityRule>>(
      endpoints.quality.ruleDetails(ruleId),
      ruleData
    );
  } catch (error) {
    logError(error, 'updateQualityRule');
    throw error;
  }
};

/**
 * Deletes a quality rule by ID
 * @param ruleId - ID of the quality rule to delete
 * @returns Promise resolving to a success indicator
 */
const deleteQualityRule = async (
  ruleId: string
): Promise<DataResponse<{ success: boolean }>> => {
  try {
    return await apiClient.delete<DataResponse<{ success: boolean }>>(
      endpoints.quality.ruleDetails(ruleId)
    );
  } catch (error) {
    logError(error, 'deleteQualityRule');
    throw error;
  }
};

/**
 * Retrieves a paginated list of quality validations with optional filtering
 * @param params - Pagination, date range, and filter parameters
 * @returns Promise resolving to a paginated list of quality validations
 */
const getQualityValidations = async (
  params: PaginationParams & DateRangeParams & { 
    executionId?: string, 
    ruleId?: string, 
    status?: ValidationStatus 
  }
): Promise<ListResponse<QualityValidation>> => {
  try {
    return await apiClient.get<ListResponse<QualityValidation>>(
      endpoints.quality.validation,
      { params }
    );
  } catch (error) {
    logError(error, 'getQualityValidations');
    throw error;
  }
};

/**
 * Retrieves a single quality validation by ID with detailed results
 * @param validationId - ID of the validation to retrieve
 * @returns Promise resolving to the detailed quality validation result
 */
const getQualityValidation = async (
  validationId: string
): Promise<DataResponse<QualityValidationResult>> => {
  try {
    return await apiClient.get<DataResponse<QualityValidationResult>>(
      endpoints.quality.validationResults(validationId)
    );
  } catch (error) {
    logError(error, 'getQualityValidation');
    throw error;
  }
};

/**
 * Retrieves a paginated list of quality issues with optional filtering
 * @param params - Pagination, date range, and filter parameters
 * @returns Promise resolving to a paginated list of quality issues
 */
const getQualityIssues = async (
  params: PaginationParams & DateRangeParams & { 
    dataset?: string, 
    table?: string, 
    dimension?: QualityDimension, 
    severity?: string,
    status?: QualityIssueStatus 
  }
): Promise<ListResponse<QualityIssue>> => {
  try {
    return await apiClient.get<ListResponse<QualityIssue>>(
      endpoints.quality.issues,
      { params }
    );
  } catch (error) {
    logError(error, 'getQualityIssues');
    throw error;
  }
};

/**
 * Retrieves a single quality issue by ID
 * @param issueId - ID of the quality issue to retrieve
 * @returns Promise resolving to the quality issue data
 */
const getQualityIssue = async (
  issueId: string
): Promise<DataResponse<QualityIssue>> => {
  try {
    return await apiClient.get<DataResponse<QualityIssue>>(
      endpoints.quality.issueDetails(issueId)
    );
  } catch (error) {
    logError(error, 'getQualityIssue');
    throw error;
  }
};

/**
 * Updates the status of a quality issue
 * @param issueId - ID of the quality issue to update
 * @param statusData - Status update data
 * @returns Promise resolving to the updated quality issue
 */
const updateQualityIssueStatus = async (
  issueId: string,
  statusData: { status: QualityIssueStatus, comment?: string }
): Promise<DataResponse<QualityIssue>> => {
  try {
    // Construct the status update endpoint
    const statusEndpoint = `${endpoints.quality.issueDetails(issueId)}/status`;
    return await apiClient.put<DataResponse<QualityIssue>>(
      statusEndpoint,
      statusData
    );
  } catch (error) {
    logError(error, 'updateQualityIssueStatus');
    throw error;
  }
};

/**
 * Retrieves quality summaries for all datasets or filtered by dataset name
 * @param params - Optional dataset filter
 * @returns Promise resolving to a list of dataset quality summaries
 */
const getDatasetQualitySummaries = async (
  params: { dataset?: string }
): Promise<ListResponse<DatasetQualitySummary>> => {
  try {
    return await apiClient.get<ListResponse<DatasetQualitySummary>>(
      `${endpoints.quality.metrics}/datasets`,
      { params }
    );
  } catch (error) {
    logError(error, 'getDatasetQualitySummaries');
    throw error;
  }
};

/**
 * Retrieves the quality score for a specific dataset and table
 * @param dataset - Dataset name
 * @param table - Table name
 * @returns Promise resolving to the quality score data
 */
const getQualityScore = async (
  dataset: string,
  table: string
): Promise<DataResponse<QualityScore>> => {
  try {
    return await apiClient.get<DataResponse<QualityScore>>(
      `${endpoints.quality.metrics}/score/${dataset}/${table}`
    );
  } catch (error) {
    logError(error, 'getQualityScore');
    throw error;
  }
};

/**
 * Retrieves quality score time series data for trend analysis
 * @param params - Dataset, optional table, and date range parameters
 * @returns Promise resolving to quality time series data
 */
const getQualityTimeSeries = async (
  params: { 
    dataset: string, 
    table?: string, 
    startDate: string, 
    endDate: string 
  }
): Promise<DataResponse<QualityTimeSeries>> => {
  try {
    return await apiClient.get<DataResponse<QualityTimeSeries>>(
      `${endpoints.quality.metrics}/timeseries`,
      { params }
    );
  } catch (error) {
    logError(error, 'getQualityTimeSeries');
    throw error;
  }
};

/**
 * Retrieves detailed quality information for a specific column
 * @param dataset - Dataset name
 * @param table - Table name
 * @param column - Column name
 * @returns Promise resolving to column quality details
 */
const getColumnQualityDetails = async (
  dataset: string,
  table: string,
  column: string
): Promise<DataResponse<ColumnQualityDetails>> => {
  try {
    return await apiClient.get<DataResponse<ColumnQualityDetails>>(
      `${endpoints.quality.metrics}/column/${dataset}/${table}/${column}`
    );
  } catch (error) {
    logError(error, 'getColumnQualityDetails');
    throw error;
  }
};

/**
 * Retrieves overall quality statistics for the system
 * @returns Promise resolving to quality statistics data
 */
const getQualityStatistics = async (): Promise<DataResponse<QualityStatistics>> => {
  try {
    return await apiClient.get<DataResponse<QualityStatistics>>(
      `${endpoints.quality.metrics}/statistics`
    );
  } catch (error) {
    logError(error, 'getQualityStatistics');
    throw error;
  }
};

/**
 * Retrieves available quality rule templates
 * @param params - Optional rule type and dimension filters
 * @returns Promise resolving to a list of quality rule templates
 */
const getQualityRuleTemplates = async (
  params: { 
    ruleType?: QualityRuleType, 
    dimension?: QualityDimension 
  }
): Promise<ListResponse<QualityRuleTemplate>> => {
  try {
    return await apiClient.get<ListResponse<QualityRuleTemplate>>(
      `${endpoints.quality.rules}/templates`,
      { params }
    );
  } catch (error) {
    logError(error, 'getQualityRuleTemplates');
    throw error;
  }
};

/**
 * Triggers an on-demand quality validation for a dataset and table
 * @param validationParams - Dataset, table, and optional rule IDs to validate
 * @returns Promise resolving to the validation ID
 */
const runQualityValidation = async (
  validationParams: { 
    dataset: string, 
    table: string, 
    ruleIds?: string[] 
  }
): Promise<DataResponse<{ validationId: string }>> => {
  try {
    return await apiClient.post<DataResponse<{ validationId: string }>>(
      `${endpoints.quality.validation}/run`,
      validationParams
    );
  } catch (error) {
    logError(error, 'runQualityValidation');
    throw error;
  }
};

// Export all functions as a single service object
export default {
  getQualityRules,
  getQualityRule,
  createQualityRule,
  updateQualityRule,
  deleteQualityRule,
  getQualityValidations,
  getQualityValidation,
  getQualityIssues,
  getQualityIssue,
  updateQualityIssueStatus,
  getDatasetQualitySummaries,
  getQualityScore,
  getQualityTimeSeries,
  getColumnQualityDetails,
  getQualityStatistics,
  getQualityRuleTemplates,
  runQualityValidation
};